### PR TITLE
Editable installs improvements

### PR DESCRIPTION
**Pull request by dnicolodi**

### CODE REVIEW

**Comment by lesteve** [comment]
Thanks for this PR!

I gave in a go to see whether that would fix the issue we were seeing in scikit-learn where some of our tests are using `pkgutil.walk_packages` as I mentioned in https://github.com/mesonbuild/meson-python/issues/557#issuecomment-1891684893.

Right now if I try to run the problematic test with pytest, I get an error `AttributeError: 'MesonpyPathFinder' object has no attribute 'find_spec'`, see below for full stack-trace. 

Let me know if I can provide additional information!

<details>
<summary>Full stack-trace</summary>

```
❯ pytest sklearn/tests/test_common.py -k are_importable
+ /home/lesteve/micromamba/envs/del/bin/ninja
ninja: no work to do.
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/pytest/__main__.py", line 5, in <module>
    raise SystemExit(pytest.console_main())
                     ^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 197, in console_main
    code = main()
           ^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 155, in main
    config = _prepareconfig(args, plugins)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 337, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/pluggy/_hooks.py", line 501, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/pluggy/_manager.py", line 119, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/pluggy/_callers.py", line 138, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/pluggy/_callers.py", line 121, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/helpconfig.py", line 104, in pytest_cmdline_parse
    config = yield
             ^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/pluggy/_callers.py", line 102, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 1094, in pytest_cmdline_parse
    self.parse(args)
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 1446, in parse
    self._preparse(args, addopts=addopts)
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 1320, in _preparse
    self.pluginmanager.consider_preparse(args, exclude_only=False)
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 727, in consider_preparse
    self.consider_pluginarg(parg)
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 753, in consider_pluginarg
    self.import_plugin(arg, consider_entry_points=True)
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/config/__init__.py", line 801, in import_plugin
    __import__(importspec)
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1322, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1262, in _find_spec
  File "/home/lesteve/micromamba/envs/del/lib/python3.12/site-packages/_pytest/assertion/rewrite.py", line 104, in find_spec
    spec = self._find_spec(name, path)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 1524, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1498, in _get_spec
AttributeError: 'MesonpyPathFinder' object has no attribute 'find_spec'
```

</details>

**Comment by lesteve** [comment]
Great, let me know if there is something I can help with!

**Comment by dnicolodi** [comment]
@lesteve Can you try again, please?

**Comment by lesteve** [comment]
Great! This seems to fix both https://github.com/mesonbuild/meson-python/issues/557#issuecomment-1891684893 and https://github.com/mesonbuild/meson-python/issues/568.

File: mesonpy/_editable.py [hunk 1]
```diff
@@ -316,6 +328,44 @@ def rebuild(self) -> Node:
...
+                continue
+            if isinstance(node, Node):
+                modname = name
+                for _, suffix in LOADERS:
+                    src = node.get('__init__' + suffix)
+                    if isinstance(src, str):
+                        ispkg = True
+            if modname and '.' not in modname:
+                yielded.add(modname)
+                yield prefix + modname, ispkg
```
Comment (rgommers) [inline mesonpy/_editable.py:365]: I've found one issue where `pkgutil.walk_packages` returns directories with files in it that get installed with `install_subdir` but don't have an `__init__.py`. According to the docs, only modules should be returned. 

It looks like the condition here should include an `is_dir` part, and if it's a directory then `ispkg` must be True for this to be yielded. Although not sure if that will mess with namespace packages; this is handled in the `find_spec` function higher up, but not in this `iter_modules`.

I ran into this in both NumPy and SciPy, where the `test_public_api.py::test_all_modules_are_expected` tests turn up a few of these non-package directories.

**Review by rgommers — COMMENTED** [review]
This is very close. I did a fair bit of testing; it's definitely a large improvement. For PyWavelets it fixed an issue, for NumPy and SciPy the tests got more accurate. There is one hiccup with `pkgutil.walk_packages`, I commented inline about what I think the problem is.

File: mesonpy/_editable.py (hunk 1 above)
Comment (dnicolodi) [reply mesonpy/_editable.py:365]: You're right. My understanding was that namespace packages should be returned too. I even added  test to check that this was the case. However, I've just tested and they are indeed not returned. Let me fix this.

Comment (dnicolodi) [reply mesonpy/_editable.py]: Thanks for the review @rgommers. The issue with namespace packages should be fixed now.

Comment (dnicolodi) [reply mesonpy/_editable.py]: The test failures are due to mesonbuild/meson#12920

Comment (rgommers) [reply mesonpy/_editable.py]: Thanks! I can confirm that it works now with both numpy and scipy, and from some more debug printing it seems to function exactly as expected.

Test failures with Meson master shouldn't be blocking here, so let's get this in.

**Review by rgommers — APPROVED** [review]

**Pull request merged**